lib/chef/provider/service/systemd.rb

```diff
@@ -76,6 +76,31 @@ def define_resource_requirements
     end
   end
 
+  def systemd_service_status
+    # If the status has already been collected, return that
+    return @service_status unless @service_status.nil?
+
+    # Collect all the status information for a service and returns it at once
+    options, args = get_systemctl_options_args
+    s = shell_out!("#{systemctl_path} #{args} show -p UnitFileState -p ActiveState #{new_resource.service_name}", options)
+    # e.g. /bin/systemctl --system show  -p UnitFileState -p ActiveState sshd.service
+    # Returns something like:
+    # ActiveState=active
+    # UnitFileState=enabled
+    status = {}
+    s.stdout.each_line do |line|
+      k, v = line.strip.split("=")
+      status[k] = v
+    end
+    # Assert requisite keys exist
+    unless status.key?("UnitFileState") && status.key?("ActiveState")
+      raise Mixlib::ShellOut::ShellCommandFailed, "'#{systemctl_path} show' not reporting status for #{new_resource.service_name}!"
+    end
+
+    @service_status = status
+    status
+  end
```

Comment:
The idiomatic ruby way to write this is memoized with a block:

```suggestion
  def systemd_service_status
    @systemd_service_status ||= begin
      # Collect all the status information for a service and returns it at once
      options, args = get_systemctl_options_args
      s = shell_out!("#{systemctl_path} #{args} show -p UnitFileState -p ActiveState #{new_resource.service_name}", options)
      # e.g. /bin/systemctl --system show  -p UnitFileState -p ActiveState sshd.service
      # Returns something like:
      # ActiveState=active
      # UnitFileState=enabled
      status = {}
      s.stdout.each_line do |line|
        k, v = line.strip.split("=")
        status[k] = v
      end
      # Assert requisite keys exist
      unless status.key?("UnitFileState") && status.key?("ActiveState")
        raise Mixlib::ShellOut::ShellCommandFailed, "'#{systemctl_path} show' not reporting status for #{new_resource.service_name}!"
      end
    end
  end
```